src/tls/ffi.c

```diff
@@ -0,0 +1,428 @@
+
+// Support temporary redirection of stdout and stderr
+typedef struct {
+    char *out_name;
+    char *err_name;
+    int fd_out_old;
+    int fd_err_old;
+    int fd_out_new;
+    int fd_err_new;
+} mitls_redirect;
+
+//
+// Initialize miTLS.
+//
+//  Called once ahead of using miTLS
+//
+//  Returns:  0 for error, nonzero for success
+//
+int  FFI_mitls_init(void)
+{
+    char *Argv[2];
+    
+    // Build a stub argv[] to satisfy caml_Startup()
+    Argv[0] = "";
+    Argv[1] = NULL;
+    
+    // Initialize the OCaml runtime
+    caml_startup(Argv);
+    
+    // Bind to functions registered via Callback.register from ML
+#define MITLS_FFI_ENTRY(x) \
+    g_mitls_FFI_##x = caml_named_value("MITLS_FFI_" # x); \
+    if (!g_mitls_FFI_##x) { \
+        printf("Failed to bind to Caml callback MITLS_FFI_" # x "\n"); \
+        return 0; \
+    }
+ MITLS_FFI_LIST  
+ #undef MITLS_FFI_ENTRY
+    
+    return 1; // success
+}
+
+void FFI_mitls_cleanup(void)
+{
+#define MITLS_FFI_ENTRY(x) \
+    g_mitls_FFI_##x = NULL;
+ MITLS_FFI_LIST  
+ #undef MITLS_FFI_ENTRY
+}
+
+char * read_stdio_file(int fd)
+{
+    struct stat st;
+    char *data;
+    
+    fstat(fd, &st);
+    // If no data was written, or a huge amount of data was written, ignore the file.
+    if (st.st_size == 0 || st.st_size > 10*1024*1024) {
+        return NULL;
+    }
+    
+    lseek(fd, SEEK_SET, 0);
+    data = malloc(st.st_size + 1); // The +1 is safe from integer overflow due to the above size check
+    if (!data) {
+        return NULL;
+    }
+    read(fd, data, st.st_size);
+    data[st.st_size] = '\0'; // Null-terminate into a C string
+    
+    return data;
+}
+
+void restore_stdio(/* in */ mitls_redirect *r, /* out */ char **outmsg, /* out */ char **errmsg)
+{
+#if REDIRECT_STDIO
+    char *msg;
+    int result;
+    
+    *outmsg = NULL;
+    *errmsg = NULL;
+ 
+    if (r->fd_out_new == -1) {
+        return;
+    }
+    
+    fflush(stdout);
+    fflush(stderr);
+    
+    // Restore stdout/stderr back to their original files
+    dup2(r->fd_out_old, STDOUT_FILENO);
+    dup2(r->fd_err_old, STDERR_FILENO);
+    
+    msg = read_stdio_file(r->fd_out_new);
+    if (msg) {
+        *outmsg = msg;
+    }
+    close(r->fd_out_new);
+    result = unlink(r->out_name);
+    if (result != 0) {
+        fprintf(stderr, "unlink %s failed errno=%d ", r->out_name, errno); perror("failed");
+    }
+    free(r->out_name);
+    
+    msg = read_stdio_file(r->fd_err_new);
+    if (msg) {
+        *errmsg = msg;
+    }
+    close(r->fd_err_new);
+    result = unlink(r->err_name);
+    if (result != 0) {
+        fprintf(stderr, "unlink %s failed errno=%d ", r->out_name, errno); perror("failed");
+    }
+    free(r->err_name);
+#else
+    r->fd_out_new = -1;
+    *outmsg = NULL;
+    *errmsg = NULL;
+#endif
+}
+
+
+void redirect_stdio(/* out */ mitls_redirect *r)
+{
+#if REDIRECT_STDIO
+    int fdout;
+    int fderr;
+    static const char template[] = "/mitlscurlXXXXXX";
```

Comment:
since we're compiling with mingw, we should use mkstemp http://man7.org/linux/man-pages/man3/mkstemp.3.html